// tsparticles/dist/Options/Classes/Particles/Opacity/OpacityAnimation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OpacityAnimation = (function () {
    function OpacityAnimation() {
        this.enable = false;
        this.minimumValue = 0;
        this.speed = 2;
        this.sync = false;
    }
    Object.defineProperty(OpacityAnimation.prototype, "opacity_min", {
        get: function () {
            return this.minimumValue;
        },
        set: function (value) {
            this.minimumValue = value;
        },
        enumerable: true,
        configurable: true
    });
    OpacityAnimation.prototype.load = function (data) {
        var _a;
        if (data !== undefined) {
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            var minimumValue = (_a = data.minimumValue) !== null && _a !== void 0 ? _a : data.opacity_min;
            if (minimumValue !== undefined) {
                this.minimumValue = minimumValue;
            }
            if (data.speed !== undefined) {
                this.speed = data.speed;
            }
            if (data.sync !== undefined) {
                this.sync = data.sync;
            }
        }
    };
    return OpacityAnimation;
}());
exports.OpacityAnimation = OpacityAnimation;
